src/lib/arns.ts

```diff
@@ -1,41 +1,144 @@
+};
+
+export type AoArNSLeaseData = AoArNSBaseNameData & {
+  type: "lease";
+  endTimestamp: number; // At what unix time (seconds since epoch) the lease ends
+};
+
+export type AoArNSPermabuyData = AoArNSBaseNameData & {
+  type: "permabuy";
+};
+export type AoArNSNameData = AoArNSPermabuyData | AoArNSLeaseData;
+export type ANTInfo = {
+  Name: string;
+  Ticker: string;
+  Denomination: number;
+  Owner: string;
+};
+
+export async function getArNSRecord(
+  name: string
+): Promise<AoArNSNameData | undefined> {
+  const ArIO = new AOProcess({ processId: AO_ARNS_PROCESS });
+  const record = await ArIO.read<AoArNSNameData>({
+    tags: [
+      { name: "Action", value: "Record" },
+      { name: "Name", value: name }
+    ]
+  });
+  return record;
 }
-const REGISTRY = "bLAgYxAdX2Ry-nt6aH2ixgvJXbpsEYm28NgJgyqfs-U";
 
-export async function getAllArNSNames(owner) {
-  const contractTxIds = await getANTsContractTxIds(owner);
-  if (contractTxIds.length === 0) return [];
+export async function getArNSRecords(): Promise<
+  Record<string, AoArNSNameData>
+> {
+  const ArIO = new AOProcess({ processId: AO_ARNS_PROCESS });
+  const record = await ArIO.read<Record<string, AoArNSNameData>>({
+    tags: [{ name: "Action", value: "Records" }]
+  });
+  return record;
+}
 
-  const url = [
-    "https://api.arns.app/v1/contract/bLAgYxAdX2Ry-nt6aH2ixgvJXbpsEYm28NgJgyqfs-U/records?",
-    ...contractTxIds.map((txId) => `contractTxId=${txId}`)
-  ].join("&");
+export async function getANTInfo(processId: string): Promise<ANTInfo> {
+  const ant = new AOProcess({ processId });
+  const tags = [{ name: "Action", value: "Info" }];
+  const info = await ant.read<ANTInfo>({ tags });
+  return info;
+}
 
-  const response = await fetch(url);
-  const data = await response.json();
-  return data;
+export async function getANTState(
+  processId: string,
+  retries = 3
+): Promise<AoANTState> {
+  const ant = new AOProcess({ processId });
+  const tags = [{ name: "Action", value: "State" }];
+  const res = await ant.read<AoANTState>({ tags, retries });
+  return res;
 }
 
+export const getAllArNSNames = async (
+  address: WalletAddress
+): Promise<string[]> => {
+  if (!address) return [];
+
+  const throttle = pLimit(50);
+
+  const arnsRecords = await getArNSRecords().then((records) =>
+    Object.values(records).filter((record) => record.processId !== undefined)
+  );
+
+  // check the contract owner and controllers
+  const results = await Promise.all(
+    arnsRecords.map(async (record) =>
+      throttle(async () => {
+        try {
+          const state = await getANTState(record.processId, 1);
+
+          if (state.Owner === address || state.Controllers.includes(address)) {
+            return state.Name;
+          }
+          return;
+        } catch (err) {
+          return;
+        }
+      })
+    )
+  );
+
+  if (results.length === 0) {
+    return [];
+  }
+
+  return Array.from(new Set(results.filter((result) => result)));
+};
```

Comment:
We need to find another way to do this (maybe collaborating with Ar.IO), or consider removing it from the contacts' info

Reply:
It has been commented out for now.